File: windows/src/RealtimeTranslator.Core/OpenAI/RealtimeTranslationMessageCodec.cs
using System;
using System.Buffers;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace RealtimeTranslator.Core.OpenAI;

/// <summary>Realtime セッションの JSON エンコード/デコード。契約は shared/protocol/endpoints.md。</summary>
public static class RealtimeTranslationMessageCodec
{
    public static byte[] Encode(RealtimeTranslationClientEvent clientEvent)
    {
        ArgumentNullException.ThrowIfNull(clientEvent);

        JsonObject payload;
        switch (clientEvent)
        {
            case RealtimeTranslationClientEvent.SessionUpdate sessionUpdate:
            {
                var config = sessionUpdate.Config;
                var input = new JsonObject();
                if (config.InputTranscriptionModel is { } model)
                {
                    input["transcription"] = new JsonObject { ["model"] = model };
                }

                // 無効時はキー省略ではなく明示的な null を送る。
                input["noise_reduction"] = config.NoiseReduction is { } noiseReduction
                    ? new JsonObject { ["type"] = noiseReduction.ToWireValue() }
                    : null;

                payload = new JsonObject
                {
                    ["type"] = "session.update",
                    ["session"] = new JsonObject
                    {
                        ["audio"] = new JsonObject
                        {
                            ["output"] = new JsonObject { ["language"] = config.OutputLanguage.ToWireValue() },
                            ["input"] = input,
                        },
                    },
                };
                break;
            }

            case RealtimeTranslationClientEvent.InputAudioBufferAppend append:
                payload = new JsonObject
                {
                    ["type"] = "session.input_audio_buffer.append",
                    ["audio"] = append.Base64Audio,
                };
                break;

            case RealtimeTranslationClientEvent.SessionClose:
                payload = new JsonObject { ["type"] = "session.close" };
                break;

            default:
                throw new ArgumentOutOfRangeException(nameof(clientEvent), clientEvent, null);
        }

        var buffer = new ArrayBufferWriter<byte>();
        using (var writer = new Utf8JsonWriter(buffer))
        {
            payload.WriteTo(writer);
        }

        return buffer.WrittenSpan.ToArray();
    }

    public static RealtimeTranslationServerEvent DecodeServerEvent(ReadOnlySpan<byte> utf8Json)
    {
        JsonNode? node;
        try
        {
            var reader = new Utf8JsonReader(utf8Json);
            node = JsonNode.Parse(ref reader);
        }
        catch (JsonException)
        {
            throw new RealtimeTranslationException(RealtimeTranslationErrorKind.InvalidMessage);
        }

        if (node is not JsonObject dictionary
            || dictionary["type"] is not JsonValue typeValue
            || !typeValue.TryGetValue<string>(out var type))
        {
            throw new RealtimeTranslationException(RealtimeTranslationErrorKind.InvalidMessage);
        }

        switch (type)
        {
            case "session.created":
                return new RealtimeTranslationServerEvent.SessionCreated();
            case "session.updated":
                return new RealtimeTranslationServerEvent.SessionUpdated();
            case "session.input_transcript.delta":
                return new RealtimeTranslationServerEvent.InputTranscriptDelta(
                    StringValue(dictionary["delta"]) ?? string.Empty,
                    StringValue(dictionary["event_id"]),
                    IntValue(dictionary["elapsed_ms"]));
            case "session.output_transcript.delta":
                return new RealtimeTranslationServerEvent.OutputTranscriptDelta(
                    StringValue(dictionary["delta"]) ?? string.Empty,
                    StringValue(dictionary["event_id"]),
                    IntValue(dictionary["elapsed_ms"]));
            case "session.output_audio.delta":
                return new RealtimeTranslationServerEvent.OutputAudioDelta();
            case "session.closed":
                return new RealtimeTranslationServerEvent.SessionClosed();
            case "error":
            {
                var errorObject = dictionary["error"] as JsonObject;
                var message = StringValue(errorObject?["message"])
                    ?? StringValue(dictionary["message"])
                    ?? RealtimeTranslationException.GenericServerMessage;
                var code = StringValue(errorObject?["code"]) ?? StringValue(errorObject?["type"]);
                return new RealtimeTranslationServerEvent.ServerError(message, code);
            }

            default:
                return new RealtimeTranslationServerEvent.Unknown(type);
        }
    }

    private static string? StringValue(JsonNode? node) =>
        node is JsonValue value && value.TryGetValue<string>(out var text) ? text : null;

    /// <summary>整数・実数どちらで届いても Int32 へ寄せる。実数は 0 方向へ切り捨てる。</summary>
    private static int? IntValue(JsonNode? node)
    {
        if (node is not JsonValue value)
        {
            return null;
        }

        if (value.TryGetValue<int>(out var intValue))
        {
            return intValue;
        }

        if (value.TryGetValue<double>(out var doubleValue)
            && doubleValue >= int.MinValue
            && doubleValue <= int.MaxValue)
        {
            return (int)doubleValue;
        }

        return null;
    }
}

File: RealtimeTranslator/OpenAI/RealtimeTranslationMessageCodec.swift
import Foundation

enum RealtimeTranslationMessageCodec {
    static func encode(_ event: RealtimeTranslationClientEvent) throws -> Data {
        let object: [String: Any]
        switch event {
        case .sessionUpdate(let config):
            var audio: [String: Any] = [
                "output": [
                    "language": config.outputLanguage.rawValue,
                ],
            ]
            var input: [String: Any] = [:]
            if let model = config.inputTranscriptionModel {
                input["transcription"] = ["model": model]
            }
            if let noiseReduction = config.noiseReduction {
                input["noise_reduction"] = ["type": noiseReduction.rawValue]
            } else {
                input["noise_reduction"] = NSNull()
            }
            if !input.isEmpty {
                audio["input"] = input
            }
            object = [
                "type": "session.update",
                "session": [
                    "audio": audio,
                ],
            ]
        case .inputAudioBufferAppend(let base64Audio):
            object = [
                "type": "session.input_audio_buffer.append",
                "audio": base64Audio,
            ]
        case .sessionClose:
            object = [
                "type": "session.close",
            ]
        }
        return try JSONSerialization.data(withJSONObject: object, options: [])
    }

    static func decodeServerEvent(from data: Data) throws -> RealtimeTranslationServerEvent {
        let object: Any
        do {
            object = try JSONSerialization.jsonObject(with: data, options: [])
        } catch {
            // shared/fixtures の decodeFailures 契約: parser 例外は invalidMessage へ正規化する。
            throw RealtimeTranslationError.invalidMessage
        }
        guard let dictionary = object as? [String: Any],
              let type = dictionary["type"] as? String
        else {
            throw RealtimeTranslationError.invalidMessage
        }

        switch type {
        case "session.created":
            return .sessionCreated
        case "session.updated":
            return .sessionUpdated
        case "session.input_transcript.delta":
            return .inputTranscriptDelta(
                delta: dictionary["delta"] as? String ?? "",
                eventID: dictionary["event_id"] as? String,
                elapsedMs: intValue(dictionary["elapsed_ms"])
            )
        case "session.output_transcript.delta":
            return .outputTranscriptDelta(
                delta: dictionary["delta"] as? String ?? "",
                eventID: dictionary["event_id"] as? String,
                elapsedMs: intValue(dictionary["elapsed_ms"])
            )
        case "session.output_audio.delta":
            // 字幕MVPでは音声payloadをデコードしない。
            return .outputAudioDelta
        case "session.closed":
            return .sessionClosed
        case "error":
            let errorObject = dictionary["error"] as? [String: Any]
            let message = (errorObject?["message"] as? String)
                ?? (dictionary["message"] as? String)
                ?? RealtimeTranslationError.genericServerMessage
            let code = (errorObject?["code"] as? String)
                ?? (errorObject?["type"] as? String)
            return .error(message: message, code: code)
        default:
            return .unknown(type: type)
        }
    }

    private static func intValue(_ value: Any?) -> Int? {
        switch value {
        case let int as Int:
            return int
        case let number as NSNumber:
            return number.intValue
        case let double as Double:
            return Int(double)
        default:
            return nil
        }
    }
}
